### PR TITLE
[Kernel] Add mocked unit tests for log segment construction

### PR DESCRIPTION
**Pull request by allisonport-db**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [X] Kernel
- [ ] Other (fill in here)

## Description

Adds mocked unit tests for the log segment construction code.

Items to address
1) [P0] For some of the _valid file listings_ the error messages are confusing / misleading. We should fix these.
2) [P2] Behavior for _corrupt file listings_ is sporadic and inconsistent and error messages are confusing. We should consider fixing these at some point.
3) [P2] The log segment building code is really confusing. I think completely re-doing this has been discussed (?) at some point but is low priority. I'm planning to revisit this PR and see what low-effort improvements I can make for readability (max 15-20 minutes.)

## How was this patch tested?

Test PR.

### CODE REVIEW

File: kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala [hunk 1]
```diff
@@ -15,12 +15,22 @@
  */
 package io.delta.kernel
```
Comment (vkorukanti) [inline kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala:16]: can you move this to internal, given it prod class is an internal one?

File: kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala [hunk 2]
```diff
@@ -93,4 +103,689 @@ class SnapshotManagerSuite extends AnyFunSuite {
...
+  //////////////////////////////////////////////////////////////////////////////////
+  // getLogSegmentForVersion tests
+  //////////////////////////////////////////////////////////////////////////////////
+
+  /* ------------------HELPER METHODS------------------ */
+
+  private val logPath = new Path("/fake/path/to/table/_delta_log")
+
+  private def deltaFileStatuses(deltaVersions: Seq[Long]): Seq[FileStatus] = {
+    assert(deltaVersions.size == deltaVersions.toSet.size)
```
Comment (vkorukanti) [inline kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala:116]: ===

File: kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala [hunk 3]
```diff
@@ -93,4 +103,689 @@ class SnapshotManagerSuite extends AnyFunSuite {
...
+    deltaVersions.map(v => FileStatus.of(FileNames.deltaFile(logPath, v), v, v))
```
Comment (vkorukanti) [inline kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala:117]: is the `v` passed to `size` and `modificationTime` intentional?

Comment (vkorukanti) [reply kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala:117]: nevermind, I  see you are using it in the tests.

File: kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala [hunk 4]
```diff
@@ -93,4 +103,689 @@ class SnapshotManagerSuite extends AnyFunSuite {
...
+
+  private def createMockFileSystemClient(
+    listFromPath: String => Seq[FileStatus]): FileSystemClient = {
+
+    new FileSystemClient {
+
+      override def listFrom(filePath: String): CloseableIterator[FileStatus] = {
+        val fileStatuses = listFromPath(filePath)
+
+        new CloseableIterator[FileStatus] {
```
Comment (vkorukanti) [inline kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala:763]: There is a utility method `toCloseableIterator`. May need to convert it to Java iterator.

File: kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala [hunk 5]
```diff
@@ -93,4 +103,689 @@ class SnapshotManagerSuite extends AnyFunSuite {
...
+    // (on first run newFiles.isEmpty() but since startingCheckpointOpt.isPresent() re-list from 0)
+    testWithSingularAndMultipartCheckpoint(
+      (0L until 25L),
+      Seq(10L, 20L),
+      startCheckpoint = Optional.of(30)
+    )
+  }
+
+  // TODO recover from missing checkpoint (getLogSegmentWithMaxExclusiveCheckpointVersion)
+  test("getLogSegmentForVersion: corrupt _last_checkpoint refers to in range version " +
```
Comment (vkorukanti) [inline kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala:657]: Does Delta-Spark handle this case where the checkpoint file referred by the last checkpoint file doesn't exist or part of it missing?

File: kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala [hunk 6]
```diff
@@ -93,4 +103,689 @@ class SnapshotManagerSuite extends AnyFunSuite {
...
+    )
+  }
+
+  test("getLogSegmentForVersion: corrupt log but reading outside corrupted range") {
+    testNoCheckpoint(
+      (0L until 5L) ++ (6L until 9L),
+      Optional.of(4)
+    )
+    testWithSingularAndMultipartCheckpoint(
+      15L until 25L,
```
Comment (vkorukanti) [inline kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala:634]: nit: adding param names here for readability.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java [hunk 7]
```diff
@@ -272,13 +272,15 @@ private Optional<LogSegment> getLogSegmentFrom(
      * @return Some LogSegment to build a Snapshot if files do exist after the given
      * startCheckpoint. None, if the directory was missing or empty.
      */
-    private Optional<LogSegment> getLogSegmentForVersion(
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java:275]: nit: can we make this protected scope, once the suite is moved to the same package.

File: kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala (hunk 5 above)
Comment (allisonport-db) [reply kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala]: yes, `getLogSegmentWithMaxExclusiveCheckpointVersion` in SnapshotManager just needs to be implemented (should be mostly a copy from the spark counter-part)

File: kernel/kernel-api/src/test/scala/io/delta/kernel/internal/SnapshotManagerSuite.scala [hunk 8]
```diff
@@ -0,0 +1,778 @@
...
+        Optional.empty(),
+        Optional.empty())
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////
+  // getLogSegmentForVersion tests
+  //////////////////////////////////////////////////////////////////////////////////
+
+  /* ------------------HELPER METHODS------------------ */
```
Comment (scottsand-db) [inline kernel/kernel-api/src/test/scala/io/delta/kernel/internal/SnapshotManagerSuite.scala:112]: are these helper methods ONLY for `getLogSegmentForVersion tests`? Or, will they be helpful for other tests? If so, they shouldn't be under the big heading

```
  //////////////////////////////////////////////////////////////////////////////////
  // getLogSegmentForVersion tests
  //////////////////////////////////////////////////////////////////////////////////
```

Comment (allisonport-db) [reply kernel/kernel-api/src/test/scala/io/delta/kernel/internal/SnapshotManagerSuite.scala:112]: The first 3 could be useful for other file listing mockings. They could probably just be in the companion object instead. Do you want me to move them or are you making changes here?

Comment (scottsand-db) [reply kernel/kernel-api/src/test/scala/io/delta/kernel/internal/SnapshotManagerSuite.scala:112]: > The first 3 could be useful for other file listing mockings. They could probably just be in the companion object instead.

Yeah, let's do that. I'm not making changes here, just browsing the code

**Pull request merged**